instance/tasks.py

```diff
@@ -339,3 +344,117 @@ def cleanup_old_betatest_users():
         users_to_delete = queryset.filter(is_active=False, profile__modified__lte=delete_cutoff)
         logger.info('Deleting %s users.', users_to_delete.count())
         users_to_delete.delete()
+
+
+class KillZombiesRunner:
+    """
+    Helper class to run `kill_zombies_periodically`
+    """
+
+    def __init__(self):
+        self.region: str = settings.OPENSTACK_REGION
+        self.recipient: list = [email for _, email in settings.ADMINS]
+        self.threshold: int = getattr(settings, "KILL_ZOMBIES_WARNING_THRESHOLD", 10)
+
+    def send_email(self, subject: str, body: str) -> int:
+        """
+        Utility method for sending emails
+        """
+        if self.recipient:
+            email = EmailMessage(
+                subject, body, settings.DEFAULT_FROM_EMAIL, self.recipient
+            )
+            email.send()
+        else:
+            logging.info("Email recipient is undefined. Skipping email.")
+
+    def trigger_warning(self):
+        """
+        Warns when more than `threshold` VMs will be terminated
+        """
+        subject = "Zombie instances are over the current threshold ({})".format(
+            self.threshold
+        )
+        body = (
+            "The number of zombie OpenStack VMs in the {} region "
+            "is over the KILL_ZOMBIES_WARNING_THRESHOLD ({}).\n"
+        ).format(self.region, self.threshold)
+        logging.info(
+            "%s\nSending an email notification to: %s",
+            body, self.recipient
+        )
+        self.send_email(subject, body)
+
+    def get_zombie_servers_count(self, stdout: str) -> int:
+        """
+        If there are servers to terminate, a `kill_zombies` dry run will warn:
+        "Would have terminated {} zombies" to stdout.
+        """
+        num_instances = 0
+        if "No servers found in region" in stdout:
+            return num_instances
+        try:
+            num_instances = int(stdout.split("Would have terminated")[1].split()[0])
+        except IndexError:
+            logger.info("Received unexpected input from dry run.")
+        return num_instances
+
+    def run(self):
+        """
+        Main method for running the task
+        """
+        try:
+            dry_run_output, output = "", ""
+            dry_run_output = call_command(
+                "kill_zombies",
+                region=self.region,
+                dry_run=True,
+            )
+            num_of_zombies = self.get_zombie_servers_count(dry_run_output)
+            if num_of_zombies > self.threshold:
+                self.trigger_warning()
+            output = call_command("kill_zombies", region=self.region)
```

Comment:
@spokerman12, we can optimize by not running the 2nd `call_command` call if the `num_of_zombies` is `0`. What do you say?

Reply:
True. Added.